### PR TITLE
[Update] Sample Viewer category front page

### PR DESCRIPTION
**Pull request by CalebRas**
## Description

This PR updates the front page of the sample viewer to organize the samples into categories, similar to the old iOS Sample Viewer. The new common design categories were used. All of the currently implemented samples fall under one of the sample categories. 

## Linked Issue(s)

- `swift/issues/4229`

## How To Test

* Load the sample viewer.
* Tap on a category to navigate to a list of samples in that category. 
* Tap on the search bar to search for a sample.

## Screenshots

|Before|After|
|:-:|:-:|
|Simulator Screen Shot - iPhone 14 Pro - 2023-07-06 at 15 49 35|Simulator Screen Shot - iPhone 14 Pro - 2023-07-06 at 15 43 26|

## To Discuss

* There is a bug where the navigation title at the top of the samples list jumps after tapping on a category and the list first loads.
* There is a bug when you back out of a sample after you entered it through a category. It then loads to the category view instead of the samples list view and it no longer lets you in that specific category.

### CODE REVIEW

File: Shared/Supporting Files/Views/CategoryView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryView.swift:31]: If the VStack is only for grouping the conditional views and no other purpose, use [`Group`](https://developer.apple.com/documentation/swiftui/group) instead

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryGridView.swift:49]: Currently if you rotate the iPhone, the grid item becomes rectangular. You can change the `aspectRatio` of `CategoryTitleView`.

Also if you tap the "corner radius" portion of an item, it still detects the tap. Try clipped modifier.

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryGridView.swift:37]:
```suggestion
        let categoriesSet = Set(samples.map(\.category))
        categories = categoriesSet.sorted()
```
When I love the expressibility and conciseness of Swift 🤩

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryGridView.swift:84]: Instead of using multiple spacers, you can hardcode an y offset to the image.
```suggestion
                        Text(category)
                            .foregroundColor(colorScheme == .dark ? .black : .white)
                            .frame(maxWidth: .infinity)
                            .offset(y: 45)
```

**Review by yo1995 — COMMENTED** [review]
I'm not particularly good at naming views, so let's discuss that with other reviewers in the next step.

First batch of comments.

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoryGridView.swift:45]: `.clipped`, `.clipShape`, and `.cornerRadius` appear to only affect the drawing and do not change the tap area of the `NavigationLink` no matter where they are applied. Do you know of a modifier that does?

File: Shared/Supporting Files/Views/CategoryView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryView.swift:32]: As a placeholder it is fine. I was thinking sth like this

```swift
    private var displayedSamples: [Sample] {
        if query.isEmpty {
            return []
        } else {
            return samples.filter { sample in
                // Can use better method names
                conditionName(sample.name, query) ||
                conditionCategory(sample.category, query) ||
                conditionTags(sample.tags, query)
            }
        }
    }
```

---

The other thing I was thinking is that probably the search result (displayedSamples) can also include the "source" of the result. Sth like 

```
Search term:

map

Result:

Display map
Display a map with imagery basemap.
matches: title, description, tags
```

This can be a follow up task in the future.

Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoryView.swift:32]: I already have a prototype for the search working, but I was saving it for a future PR. Is this placeholder good enough for this PR, or do you want me to update it?

Comment (yo1995) [reply Shared/Supporting Files/Views/CategoryView.swift:32]: Yes, let's do it in another PR.

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryGridView.swift:81]: There are 2 empty lines at the end of this file. Please make it 1.

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (yo1995) [inline Shared/Supporting Files/Views/CategoryGridView.swift:74]: Can this frame modifier be removed? Does it have some purpose 🤔

**Review by yo1995 — COMMENTED** [review]
LGTM

**Comment by yo1995** [comment]
I added David to the reviewers as he is familiar with UI best practices.

**Comment by yo1995** [comment]
> There is a bug when you back out of a sample after you entered it through a category. It then loads to the category view instead of the samples list view and it no longer lets you in that specific category.

I read the code and couldn't find anything that might caused this bug, except the difference between iOS 16 `NavigationSplitView` and iOS 15 `NavigationView`. I also tested on iOS 15.5 and the bug is not there 🤔 . @philium If you have a chance can you help take a look at `CategoryGridView` and see if anything stands out? Thank you.

**Comment by des12437** [comment]
When selecting a category and then a sample, I notice that the back button in the sample will navigate back to the categories view rather than the selected category view. This seems like unexpected behavior since a navigation link should push/pop one view at a time.


https://github.com/Esri/arcgis-maps-sdk-swift-samples/assets/117859673/fcd1b8b3-8e56-414c-8948-6aa951c4b51c

File: Shared/Assets.xcassets/Categories/Cloud-and-Portal-bg.imageset/Cloud-and-Portal-bg@2x.png
Comment (CalebRas) [reply Shared/Assets.xcassets/Categories/Cloud-and-Portal-bg.imageset/Cloud-and-Portal-bg@2x.png:1]: Yes. The images still pertain to a category, there are just no samples that we have implemented in that category yet. Categories with no samples are hidden, which was a suggestion from Divesh. The categories are generated from the samples, so as soon as one with an Augmented Reality sample is implemented, that category will show and the images will be used.

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (CalebRas) [reply Shared/Supporting Files/Views/CategoryGridView.swift:56]: I'll keep the extension since the same pattern was used by others, like in the `SampleListView`.

File: Shared/Supporting Files/Views/CategoryView.swift
Comment (des12437) [inline Shared/Supporting Files/Views/CategoryView.swift:47]:
```suggestion
                    isAboutViewPresented = true
```

**Review by des12437 — COMMENTED** [review]
Looks good. I'm not sure if the `AboutView` is within the scope of this PR, but noticed that the logo and copyright text could use some spacing.




The copyright year can also be generated rather than hardcoded:

```
Text("Copyright © 2022 - \(String(Calendar.current.component(.year, from: Date()))) Esri. All Rights Reserved.")
```

File: Shared/Supporting Files/Views/CategoryGridView.swift
Comment (dfeinzimer) [inline Shared/Supporting Files/Views/CategoryGridView.swift:48]:
```suggestion
                    .contentShape(RoundedRectangle(cornerRadius: 30))
                    .buttonStyle(.plain)
                }
```

Mac Catalyst before:
Screenshot 2023-07-17 at 18 15 37

After:              
Screenshot 2023-07-17 at 18 18 04

**Comment by CalebRas** [comment]
> Looks good. I'm not sure if the `AboutView` is within the scope of this PR, but noticed that the logo and copyright text could use some spacing.
> 
> 
> The copyright year can also be generated rather than hardcoded:
> 
> ```
> Text("Copyright © 2022 - \(String(Calendar.current.component(.year, from: Date()))) Esri. All Rights Reserved.")
> ```

Yea I think this would fall under another issue/PR.

**Review by dfeinzimer — APPROVED** [review]

**Comment by yo1995** [comment]
Please see Phil's comment on the second part of the year string: https://github.com/Esri/arcgis-maps-sdk-swift-samples/pull/3#discussion_r878596415 It shouldn't be dynamic, but rather manually updated to reflect the last change to the code.

**Review by des12437 — APPROVED** [review]

**Comment by yo1995** [comment]
Sorry, I meant to hold this PR from being merged until the navigation bug (mentioned in the "To Discuss" section in the description) is addressed. I'll let Phil to comment on that. 

@dfeinzimer and @des12437 do you have thought on that?

**Comment by dfeinzimer** [comment]
> @dfeinzimer and @des12437 do you have thought on that?

Hm, sorry I missed those in the description!

Neither of them seem to still be present? Am I missing?

First issue:

No jumping spotted.

Second issue:

I'm able to re-open the Analysis category.

**Comment by yo1995** [comment]
@dfeinzimer That's intriguing behavior! 😂 May I ask which Xcode and Sim version are you using? It differs from mine both on iOS 15 and 16…

> That's unless you mean the jumping the second time around when coming back into the Analysis category?

When you tap "Back" from a sample, it directly jump back to the category grid view, which means sth's wrong with the navigation stack. 🤔

**Comment by dfeinzimer** [comment]
> May I ask which Xcode and Sim version are you using? It differs from mine both on iOS 15 and 16…

Xcode 14.3.1 with a 14 Pro Max sim on iOS 17.0

**Comment by des12437** [comment]
> @dfeinzimer and @des12437 do you have thought on that?

I see the same navigation issue on Xcode 14.3.1 and iPhone 14 Pro Max simulator iOS 16.4.

**Pull request merged**
